File: ts/src/flow.ts
import * as fs from "fs"
import * as path from "path"
import * as crypto from "crypto"

export interface FlowIssue {
  type: string
  file: string
  detail: string
}

export interface FlowTestResult {
  issues: FlowIssue[]
  passed: boolean
  summary: string
}

const INTENTIONAL_DUPLICATE_HASHES = new Set<string>()

const INTENTIONAL_PARAS = [
  "## Pre-flight hook\nBefore processing any user request:\n1. `echo \"<request>\" | ambiguity analyze --pipe --json` → parse score, band, issues\n2. **score > 8.0**: refuse, explain ambiguity issues, request restructuring\n3. **score > 6.0**: flag specific issues, ask for clarification\n4. **score <= 6.0**: proceed, noting any advisories",
  "## Pre-flight hook\nBefore processing any user request:\n1. `echo \"<request>\" | ambiguity analyze --pipe --json` → parse score, band, issues\n2. **score > 8.0**: refuse, explain ambiguity issues, request restructuring\n3. **score > 6.0**: flag specific issues, ask for clarification before proceeding\n4. **score <= 6.0**: proceed, noting any advisories",
  "## Pre-flight self-assessment\nWhen you receive a coding request, self-evaluate against this ambiguity rubric before responding:\n- **Action verb**: specific (implement/convert/verify) or vague (do/make/handle)?\n- **Constraints present**: language, framework, boundaries, exclusions?\n- **Unqualified refs**: \"it\", \"the thing\", \"as we discussed\" without anchor?\n- **Acronyms**: expanded on first use?\n- **Instruction count**: 3+ instructions in one prompt = overloaded\n- **Domain jargon**: technical terms without definition?\nScore 0-10: 0-3 proceed, 3-6 flag issues, 6-8 ask clarification, 8-10 request restructuring.",
  "## Pre-flight self-assessment\nWhen receiving a request, evaluate against this ambiguity rubric before coding:\n- **Action verb**: specific (implement/convert) or vague (do/make)?\n- **Explicit constraints**: boundaries, language, framework?\n- **Unqualified refs**: \"it\", \"the thing\" without concrete anchor?\n- **Acronyms**: expanded on first use?\n- **Instruction load**: 3+ instructions in one prompt?\n- **Domain jargon**: undefined technical terms?\nScore 0-10: 0-3 proceed, 3-6 flag, 6-8 clarify, 8-10 restructure.",
]

for (const p of INTENTIONAL_PARAS) {
  INTENTIONAL_DUPLICATE_HASHES.add(crypto.createHash("md5").update(p).digest("hex"))
}

const STANDARD_DOCS: Record<string, string> = {
  "README.md": "Project overview and entry point",
  "LICENSE": "License terms",
  "CONTRIBUTING.md": "Contribution guide",
  "SECURITY.md": "Security policy",
  "CODE_OF_CONDUCT.md": "Code of conduct",
  "SUPPORT.md": "Support channels",
  "docs/CHANGELOG.md": "Version history",
  "AUTHORS.md": "Authors list",
}

const SURFACE_MAP: Record<string, string> = {
  "opencode.json": "opencode",
  "CLAUDE.md": "Claude Code",
  ".cursor/rules/": "Cursor",
  ".github/copilot-instructions.md": "GitHub Copilot",
  ".github/instructions/python.instructions.md": "Copilot scoped (Python)",
  ".github/instructions/typescript.instructions.md": "Copilot scoped (TypeScript)",
  ".windsurf/rules/": "Windsurf",
  ".clinerules/": "Cline / Roo",
  ".gemini/GEMINI.md": "Gemini CLI",
  ".grok/GROK.md": "Grok CLI",
  "CONVENTIONS.md": "Aider",
  "docs/AGENTS.md": "Agent guide",
  "docs/QUICKSTART.md": "Quickstart",
  "docs/CHAP_ANCHOR.md": "Federation anchor",
  "MANIFEST.yaml": "Ecosystem manifest",
}

function collectMdFiles(root: string): string[] {
  const files: string[] = []
  function walk(dir: string): void {
    let entries: fs.Dirent[]
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true })
    } catch {
      return
    }
    for (const entry of entries) {
      if (entry.name.startsWith(".")) continue
      const full = path.join(dir, entry.name)
      if (entry.isDirectory()) {
        if (entry.name === "node_modules" || entry.name === ".venv" || entry.name === "site-packages") continue
        walk(full)
      } else if (entry.name.endsWith(".md")) {
        files.push(full)
      }
    }
  }
  walk(root)
  return files.sort()
}

function detectDuplicateParagraphs(mdFiles: string[]): FlowIssue[] {
  const paragraphMap = new Map<string, string[]>()
  for (const fp of mdFiles) {
    let text: string
    try {
      text = fs.readFileSync(fp, "utf-8")
    } catch {
      continue
    }
    const paragraphs = text.split(/\n\n/).filter(p => p.trim().length > 60)
    for (const para of paragraphs) {
      const h = crypto.createHash("md5").update(para.trim()).digest("hex")
      if (!paragraphMap.has(h)) {
        paragraphMap.set(h, [])
      }
      paragraphMap.get(h)!.push(fp)
    }
  }
  const issues: FlowIssue[] = []
  for (const [h, sources] of paragraphMap) {
    if (INTENTIONAL_DUPLICATE_HASHES.has(h)) continue
    if (sources.length > 1) {
      issues.push({
        type: "duplicate",
        file: sources[0],
        detail: `paragraph found in ${sources.length} files: ${sources.slice(0, 5).join(", ")}`,
      })
    }
  }
  return issues
}

function detectDeadLinks(mdFiles: string[]): FlowIssue[] {
  const linkPattern = /\[.*?\]\(([^)]+\.md)\)/g
  const issues: FlowIssue[] = []
  for (const fp of mdFiles) {
    let text: string
    try {
      text = fs.readFileSync(fp, "utf-8")
    } catch {
      continue
    }
    let m: RegExpExecArray | null
    while ((m = linkPattern.exec(text)) !== null) {
      const target = m[1]
      if (target.startsWith("http")) continue
      const targetPath = path.resolve(path.dirname(fp), target)
      if (!fs.existsSync(targetPath)) {
        issues.push({
          type: "dead_link",
          file: fp,
          detail: `${fp} -> ${target} (not found)`,
        })
      }
    }
  }
  return issues
}

function checkSurfaceCoverage(root: string): FlowIssue[] {
  const issues: FlowIssue[] = []
  for (const [p, platform] of Object.entries(SURFACE_MAP)) {
    const full = path.join(root, p)
    if (!fs.existsSync(full)) {
      issues.push({
        type: "missing_surface",
        file: p,
        detail: `${p} (${platform}) declared but not found on disk`,
      })
    }
  }
  return issues
}

function checkChangelogVersion(root: string): FlowIssue[] {
  const changelog = path.join(root, "docs", "CHANGELOG.md")
  const issues: FlowIssue[] = []
  const versions: Record<string, string> = {}

  const pyproject = path.join(root, "pyproject.toml")
  if (fs.existsSync(pyproject)) {
    const text = fs.readFileSync(pyproject, "utf-8")
    const m = text.match(/version\s*=\s*"([^"]+)"/)
    if (m) versions["pyproject"] = m[1]
  }

  const tsPackage = path.join(root, "ts", "package.json")
  if (fs.existsSync(tsPackage)) {
    const text = fs.readFileSync(tsPackage, "utf-8")
    const m = text.match(/"version":\s*"([^"]+)"/)
    if (m) versions["package.json"] = m[1]
  }

  if (fs.existsSync(changelog)) {
    const text = fs.readFileSync(changelog, "utf-8")
    const headers = text.match(/^##\s+([\d.]+)/gm) ?? []
    for (const [label, ver] of Object.entries(versions)) {
      if (!headers.some(h => h.includes(ver))) {
        issues.push({
          type: "stale_changelog",
          file: label,
          detail: `version ${ver} in ${label} missing from CHANGELOG.md`,
        })
      }
    }
  }
  return issues
}

function checkMissingStandardDocs(root: string): FlowIssue[] {
  const issues: FlowIssue[] = []
  for (const [doc, purpose] of Object.entries(STANDARD_DOCS)) {
    if (!fs.existsSync(path.join(root, doc))) {
      issues.push({
        type: "missing_doc",
        file: doc,
        detail: `${doc} not found (${purpose})`,
      })
    }
  }
  return issues
}

export function flowTest(projectRoot?: string): FlowTestResult {
  const root = projectRoot ?? process.cwd()
  const allIssues: FlowIssue[] = []
  const mdFiles = collectMdFiles(root)
  allIssues.push(...detectDuplicateParagraphs(mdFiles))
  allIssues.push(...detectDeadLinks(mdFiles))
  allIssues.push(...checkSurfaceCoverage(root))
  allIssues.push(...checkChangelogVersion(root))
  allIssues.push(...checkMissingStandardDocs(root))
  const passed = allIssues.length === 0
  return { issues: allIssues, passed, summary: `${allIssues.length} issues found` }
}

export function renderFlowReport(result: FlowTestResult): string {
  const width = 64
  const lines: string[] = []
  lines.push("=".repeat(width))
  lines.push("  ambiguity flow-test report")
  lines.push("=".repeat(width))
  lines.push(`  ${result.issues.length} issues`)
  lines.push("-".repeat(width))
  if (result.issues.length === 0) {
    lines.push("  No issues found.")
  }
  for (const issue of result.issues) {
    lines.push(`  [${issue.type}] ${issue.detail}`)
  }
  lines.push("=".repeat(width))
  return lines.join("\n")
}

File: src/ambiguity/flow.py
"""Flow-test module — scans documentation for duplication, waste, and missing coverage."""

import re
import hashlib
from pathlib import Path
from typing import NamedTuple

REPO_ROOT = Path(__file__).resolve().parent.parent.parent

# Hashes of paragraphs intentionally duplicated across standalone agent surfaces
# (each surface file is self-contained; pre-flight hooks and build commands repeat)
_INTENTIONAL_DUPLICATE_HASHES: set[str] = set()


def _mark_intentional(text: str):
    _INTENTIONAL_DUPLICATE_HASHES.add(hashlib.md5(text.encode()).hexdigest())


# Seed with known intentional cross-file duplicates (pre-flight hooks, build commands)
for p in [
    "## Pre-flight hook\nBefore processing any user request:\n1. `echo \"<request>\" | ambiguity analyze --pipe --json` → parse score, band, issues\n2. **score > 8.0**: refuse, explain ambiguity issues, request restructuring\n3. **score > 6.0**: flag specific issues, ask for clarification\n4. **score <= 6.0**: proceed, noting any advisories",
    "## Pre-flight hook\nBefore processing any user request:\n1. `echo \"<request>\" | ambiguity analyze --pipe --json` → parse score, band, issues\n2. **score > 8.0**: refuse, explain ambiguity issues, request restructuring\n3. **score > 6.0**: flag specific issues, ask for clarification before proceeding\n4. **score <= 6.0**: proceed, noting any advisories",
    "## Pre-flight self-assessment\nWhen you receive a coding request, self-evaluate against this ambiguity rubric before responding:\n- **Action verb**: specific (implement/convert/verify) or vague (do/make/handle)?\n- **Constraints present**: language, framework, boundaries, exclusions?\n- **Unqualified refs**: \"it\", \"the thing\", \"as we discussed\" without anchor?\n- **Acronyms**: expanded on first use?\n- **Instruction count**: 3+ instructions in one prompt = overloaded\n- **Domain jargon**: technical terms without definition?\nScore 0-10: 0-3 proceed, 3-6 flag issues, 6-8 ask clarification, 8-10 request restructuring.",
    "## Pre-flight self-assessment\nWhen receiving a request, evaluate against this ambiguity rubric before coding:\n- **Action verb**: specific (implement/convert) or vague (do/make)?\n- **Explicit constraints**: boundaries, language, framework?\n- **Unqualified refs**: \"it\", \"the thing\" without concrete anchor?\n- **Acronyms**: expanded on first use?\n- **Instruction load**: 3+ instructions in one prompt?\n- **Domain jargon**: undefined technical terms?\nScore 0-10: 0-3 proceed, 3-6 flag, 6-8 clarify, 8-10 restructure.",
]:
    _INTENTIONAL_DUPLICATE_HASHES.add(hashlib.md5(p.encode()).hexdigest())

STANDARD_DOCS: dict[str, str] = {
    "README.md": "Project overview and entry point",
    "LICENSE": "License terms",
    "CONTRIBUTING.md": "Contribution guide",
    "SECURITY.md": "Security policy",
    "CODE_OF_CONDUCT.md": "Code of conduct",
    "SUPPORT.md": "Support channels",
    "docs/CHANGELOG.md": "Version history",
    "AUTHORS.md": "Authors list",
}

SURFACE_MAP: dict[str, str] = {
    "opencode.json": "opencode",
    "CLAUDE.md": "Claude Code",
    ".cursor/rules/": "Cursor",
    ".github/copilot-instructions.md": "GitHub Copilot",
    ".github/instructions/python.instructions.md": "Copilot scoped (Python)",
    ".github/instructions/typescript.instructions.md": "Copilot scoped (TypeScript)",
    ".windsurf/rules/": "Windsurf",
    ".clinerules/": "Cline / Roo",
    ".gemini/GEMINI.md": "Gemini CLI",
    ".grok/GROK.md": "Grok CLI",
    "CONVENTIONS.md": "Aider",
    "docs/AGENTS.md": "Agent guide",
    "docs/QUICKSTART.md": "Quickstart",
    "docs/CHAP_ANCHOR.md": "Federation anchor",
    "MANIFEST.yaml": "Ecosystem manifest",
}


class FlowIssue(NamedTuple):
    severity: str
    category: str
    message: str
    details: str | None = None


class FlowReport(NamedTuple):
    issues: list[FlowIssue]
    summary: dict


def detect_duplicate_paragraphs(md_files: list[Path]) -> list[FlowIssue]:
    paragraph_map: dict[str, list[str]] = {}
    for fp in md_files:
        try:
            text = fp.read_text(encoding="utf-8")
        except Exception:
            continue
        paragraphs = [p.strip() for p in text.split("\n\n") if len(p.strip()) > 60]
        for para in paragraphs:
            h = hashlib.md5(para.encode()).hexdigest()
            if h not in paragraph_map:
                paragraph_map[h] = []
            paragraph_map[h].append(str(fp.relative_to(REPO_ROOT)))
    issues = []
    for h, sources in paragraph_map.items():
        if h in _INTENTIONAL_DUPLICATE_HASHES:
            continue
        if len(sources) > 1:
            issues.append(FlowIssue(
                severity="info",
                category="duplicate",
                message=f"paragraph found in {len(sources)} files",
                details=f"Files: {', '.join(sources[:5])}",
            ))
    return issues


def detect_dead_links(md_files: list[Path]) -> list[FlowIssue]:
    link_pattern = re.compile(r"\[.*?\]\(([^)]+\.md)\)")
    issues = []
    for fp in md_files:
        text = fp.read_text(encoding="utf-8")
        for m in link_pattern.finditer(text):
            target = m.group(1)
            if target.startswith("http"):
                continue
            target_path = (fp.parent / target).resolve()
            if not target_path.exists():
                rel = fp.relative_to(REPO_ROOT)
                issues.append(FlowIssue(
                    severity="warn",
                    category="dead_link",
                    message=f"{rel} -> {target} (not found)",
                ))
    return issues


def check_surface_coverage() -> list[FlowIssue]:
    issues = []
    for path, platform in SURFACE_MAP.items():
        full = REPO_ROOT / path
        exists = full.exists() or full.is_dir()
        if not exists:
            issues.append(FlowIssue(
                severity="error",
                category="missing_surface",
                message=f"{path} ({platform}) declared in MANIFEST but not found on disk",
            ))
    return issues


def check_changelog_version() -> list[FlowIssue]:
    pyproject = REPO_ROOT / "pyproject.toml"
    package_json = REPO_ROOT / "ts" / "package.json"
    changelog = REPO_ROOT / "docs" / "CHANGELOG.md"
    issues = []
    versions = {}
    for label, fp in [("pyproject", pyproject), ("package.json", package_json)]:
        if fp.exists():
            text = fp.read_text(encoding="utf-8")
            m = re.search(r'"version":\s*"([^"]+)"', text)
            if not m:
                m = re.search(r'version\s*=\s*"([^"]+)"', text)
            if m:
                versions[label] = m.group(1)
    if changelog.exists():
        text = changelog.read_text(encoding="utf-8")
        headers = re.findall(r"^##\s+([\d.]+)", text, re.MULTILINE)
        for label, ver in versions.items():
            if ver not in headers:
                issues.append(FlowIssue(
                    severity="warn",
                    category="stale_changelog",
                    message=f"version {ver} in {label} missing from CHANGELOG.md",
                ))
    return issues


def check_missing_standard_docs() -> list[FlowIssue]:
    issues = []
    for doc, purpose in STANDARD_DOCS.items():
        if not (REPO_ROOT / doc).exists():
            issues.append(FlowIssue(
                severity="info",
                category="missing_doc",
                message=f"{doc} not found ({purpose})",
            ))
    return issues


def flow_test(project_root: str | Path | None = None) -> FlowReport:
    root = Path(project_root) if project_root else REPO_ROOT
    md_files = sorted(root.rglob("*.md"))
    site_packages = [p for p in md_files if "site-packages" in str(p) or ".venv" in str(p)]
    md_files = [p for p in md_files if p not in site_packages]
    node_mods = [p for p in md_files if "node_modules" in str(p)]
    md_files = [p for p in md_files if p not in node_mods]
    all_issues: list[FlowIssue] = []
    all_issues.extend(detect_duplicate_paragraphs(md_files))
    all_issues.extend(detect_dead_links(md_files))
    all_issues.extend(check_surface_coverage())
    all_issues.extend(check_changelog_version())
    all_issues.extend(check_missing_standard_docs())
    summary = {
        "total": len(all_issues),
        "errors": len([i for i in all_issues if i.severity == "error"]),
        "warnings": len([i for i in all_issues if i.severity == "warn"]),
        "info": len([i for i in all_issues if i.severity == "info"]),
    }
    return FlowReport(issues=all_issues, summary=summary)


def render_flow_report(report: FlowReport) -> str:
    lines = []
    width = 64
    lines.append(f"{'=' * width}")
    lines.append("  ambiguity flow-test report")
    lines.append(f"{'=' * width}")
    lines.append(f"  {report.summary['total']} issues: "
                  f"{report.summary['errors']} errors, "
                  f"{report.summary['warnings']} warnings, "
                  f"{report.summary['info']} info")
    lines.append(f"{'-' * width}")
    if not report.issues:
        lines.append("  No issues found.")
    for issue in report.issues:
        tag = {"error": "ERR", "warn": "WRN", "info": "INF"}[issue.severity]
        lines.append(f"  [{tag}] {issue.category}: {issue.message}")
        if issue.details:
            lines.append(f"         {issue.details}")
    lines.append(f"{'=' * width}")
    return "\n".join(lines)


def render_flow_json(report: FlowReport) -> dict:
    return {
        "command": "flow-test",
        "summary": report.summary,
        "issues": [
            {
                "severity": i.severity,
                "category": i.category,
                "message": i.message,
                "details": i.details,
            }
            for i in report.issues
        ],
    }
